Update README (#32)
elasticio/elasticio-cli: README.md

# Elasticio CLI Tool
## Table of Contents
- [Installation](#Installation)
- [Creating Test Fixtures](#Creating-Test-Fixtures)
- [Commands](#Commands)
  - [cmp:process](#cmp:process)
  - [cmp:exec](#cmp:exec)
  - [cmp:validate](#cmp:validate)
- [Other Information](#Other-Information)

## Installation

> To install this tool, run:

Reviewer: Can you add remarks about how to set up auto-completion? https://www.npmjs.com/package/caporal#shell-auto-completion

Related: it appears as though the `:` character breaks auto-completion.

Author: I am getting an error every time I try to set up autocompletion myself, that looks like the below.
Would it make more sense to leave the autocomplete to another issue for now since this is time-sensitive? I don't think this error is from our side, since I followed the caporal directions directly, but I am not sure?

```
events.js:187
      throw er; // Unhandled 'error' event
      ^

Error: EPIPE: broken pipe, write
    at Socket._write (internal/net.js:58:25)
    at doWrite (_stream_writable.js:431:12)
    at writeOrBuffer (_stream_writable.js:415:5)
    at Socket.Writable.write (_stream_writable.js:305:11)
    at CaporalTransport.log (C:\Users\Hannah Bulmer\elasticio-cli\node_modules\caporal\lib\logger.js:21:18)
    at transportLog (C:\Users\Hannah Bulmer\elasticio-cli\node_modules\winston\lib\winston\logger.js:234:15)
    at C:\Users\Hannah Bulmer\elasticio-cli\node_modules\async\lib\async.js:157:13
    at _each (C:\Users\Hannah Bulmer\elasticio-cli\node_modules\async\lib\async.js:57:9)
    at Object.async.each (C:\Users\Hannah Bulmer\elasticio-cli\node_modules\async\lib\async.js:156:9)
    at exports.Logger.Logger.log (C:\Users\Hannah Bulmer\elasticio-cli\node_modules\winston\lib\winston\logger.js:246:9)
    at exports.Logger.target.<computed> [as info] (C:\Users\Hannah Bulmer\elasticio-cli\node_modules\winston\lib\winston\common.js:54:18)
    at Program._handleCompletionCommand (C:\Users\Hannah Bulmer\elasticio-cli\node_modules\caporal\lib\program.js:262:21)
    at Program.parse (C:\Users\Hannah Bulmer\elasticio-cli\node_modules\caporal\lib\program.js:281:19)
    at Object.<anonymous> (C:\Users\Hannah Bulmer\elasticio-cli\elasticio:45:9)
    at Module._compile (internal/modules/cjs/loader.js:956:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
Emitted 'error' event on Socket instance at:
    at errorOrDestroy (internal/streams/destroy.js:108:12)
    at onwriteError (_stream_writable.js:446:5)
    at onwrite (_stream_writable.js:473:5)
    at Socket._write (internal/net.js:62:14)
    at doWrite (_stream_writable.js:431:12)
    [... lines matching original stack trace ...]
    at Program.parse (C:\Users\Hannah Bulmer\elasticio-cli\node_modules\caporal\lib\program.js:281:19) {
  errno: 'EPIPE',
  syscall: 'write',
  code: 'EPIPE'
}
```

Reviewer: Sure.  We can write up another issue.  The only thing that we may want to change at the moment is to remove the `:` from the commands since that appears to create issues.

     -v, --verbose      Verbose mode - will also output debug messages
````

This confirms a successful installation.

## Creating Test Fixtures

> A **test fixture** is a test sample that a piece of code should be run against, in JSON format. Every test fixture for the elasticio platform should have a `msg` field and a `cfg` field, at minimum. A `snapshot` field can also be added and used. All test fixtures are stored together in one JSON file. This should be stored in a folder called `test/fixture.json` file, where `test` is a folder stored at the same level as `component.json`.

Reviewer 2: ```suggestion
A **test fixture** is a JSON test sample that a piece of code should be run against. Every test fixture for the elasticio platform should have a `msg` field and a `cfg` field at minimum. A `snapshot` field can also be included. All test fixtures are stored together in one JSON file, `test/fixture.json`, where `test` is a folder stored at the same level as `component.json`.
```

exports.shutdown = function shutdown(cfg, startupData) { };
```

> Furthermore, the provided fixture will be initially be checked against its respective `schema`, if the action/trigger has a static schema, for invalid message. This will not impede the running of the `process` action.

Reviewer 2: ```suggestion
Furthermore, if the action/trigger has a static schema, the provided fixture will be initially be checked against its respective `schema` for an invalid message. This will not impede the running of the `process` action.
```

exports.shutdown = function shutdown(cfg, startupData) { };
```

Furthermore, the provided fixture will be initially be checked against its respective `schema`, if the action/trigger has a static schema, for invalid message. This will not impede the running of the `process` action.

The CLI tool will print all details emitted from the `process`, and any values that have been returned before it exits.

### Flags

`cmp:process` takes two optional flags:
- -x, --fixture: fixture name to run against (optional)
- -a, --action: name of action/trigger to run (optional)

To view the help menu for this function, type `elasticio cmp:process -h`.

### Examples

`> elasticio cmp:process -x success -a lookupObject`

### Limitations
None at the moment

## cmp:exec

> `cmp:exec` will allow you to run any exported function from an action/trigger that may be used on the platform, or `verifyCredentials` for a given component. For example, this allows you to run the functions `getMetaModel`, and other Select View functions. It can be run from within a component directory, or from outside by providing a path.

Reviewer 2: ```suggestion
`cmp:exec` will allow you to run `verifyCredentials` and any exported function from an action/trigger that may be used on the platform. For example, this allows you to run the functions `getMetaModel` and other Select View functions. It can be run from within a component directory, or from outside by providing a path.
```

The CLI tool assumes that exported functions and `verifyCredentials` will have the following signatures:

```javascript
exports.verify = async function verify(cfg, optional callback); // found in verifyCredentials.js
exports.getMetaModel = function getMetaModel(cfg);
exports.selectViewFunctions = function selectView(cfg);
exports.process = function process(msg, cfg, snapshot);
```

> `getMetaModel` and any Select View function should take the `cfg` as first parameter, as should `verify`. For this function, running `process` will not run it with the startup/shutdown functions, and will run only the selected method in isolation. This can be beneficial for testing purposes.

Reviewer 2: ```suggestion
`verify`, `getMetaModel`, and any Select View function should take the `cfg` as first parameter. For this function, running `process` will not run it with the startup/shutdown functions, and will run only the selected method in isolation. This can be beneficial for testing purposes.
```
